File: person_store.py
'''
a very basic store for people objects
'''
import json
from os import path


class PersonStore(object):

  def __init__(self, store='people.json'):
    self.__people = []
    self.__store  = store
    self.__load_data()


  def __load_data(self):
    if path.exists(self.__store):
      with open(self.__store, 'r') as f:
        self.__people = json.loads(f.read())


  def __save_data(self):
    with open(self.__store, 'w') as f:
      self.__people = f.write(json.dumps(self.__people))


  def __next_id(self):
    if len(self.__people) > 0:
      return max([int(p['id']) for p in self.__people]) + 1
    else:
      return 1


  def seed(self, people):
    self.__people = people
    self.__save_data()


  @property
  def people(self):
    return self.__people


  def read_person(self, person_id):
    for person in self.people:
      if person['id'] == person_id:
        return person

    return None


  def create_person(self, person):
    new_id = self.__next_id()
    person['id'] = new_id
    self.__people.append(person)
    self.__save_data()
    return person


  def update_person(self, person_id, person_data):
    for i, person in enumerate(self.people):
      if person['id'] == person_id:
        del(self.__people[i])
        self.__people.append(person_data)
        self.__save_data()
        return person

    return {}


  def delete_person(self, person_id):
    for i, person in enumerate(self.people):
      if person['id'] == person_id:
        del(self.people[i])
        self.__save_data()
        return True

    return False

File: sinatra/person_store.rb
require 'json'


class PersonStore

  attr_reader :people

  def initialize(store='people.json')
    @people = []
    @store = store
    load_data
  end


  def seed(people)
    @people = people
    save_data
  end


  def read_person(person_id)
    @people.each do |person|
      return person if person['id'] == person_id
    end

    return nil
  end


  def update_person(person_id, person_data)
    @people.each do |person|
      if person['id'] == person_id
        person_data['id'] = person_id
        @people -= [person]
        @people << person_data
        save_data
        return person_data
      end
    end

    nil
  end


  def delete_person(person_id)
    @people.each do |person|
      if person['id'] == person_id
        @people -= [person]
        save_data
        return true
      end
    end

    false
  end


  def create_person(person_data)
    person_data['id'] = next_id
    @people << person_data
    save_data
    person_data
  end


  private

  def load_data()
    if File.exists?(@store)
      data = File.read(@store)
      @people = JSON.parse(data)
    end
  end


  def save_data()
    File.open(@store, 'w') do |file|
      file.write(JSON.dump(@people))
    end
  end


  def next_id()
    if @people.length > 0
      return @people.map {|m| m['id']}.max + 1
    else
      return 1
    end
  end

end

